src/test/java/org/sagebionetworks/workers/util/semaphore/WriteReadSemaphoreImplTest.java

```diff
@@ -0,0 +1,150 @@
+package org.sagebionetworks.workers.util.semaphore;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.Optional;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.sagebionetworks.common.util.progress.ProgressCallback;
+import org.sagebionetworks.common.util.progress.ProgressListener;
+import org.sagebionetworks.database.semaphore.CountingSemaphore;
+
+@ExtendWith(MockitoExtension.class)
+public class WriteReadSemaphoreImplTest {
+
+	@Mock
+	private ProgressCallback mockCallback;
+	@Mock
+	private CountingSemaphore mockCountingSemaphore;
+	@Captor
+	private ArgumentCaptor<ProgressListener> listenerCaptor;
+
+	private WriteReadSemaphoreImpl semaphore;
+
+	private int maxNumberOfReaders;
+	private String[] keys;
+	private long maxTimeout;
+	private String context;
+
+	@BeforeEach
+	public void before() {
+		maxNumberOfReaders = 4;
+		keys = new String[] { "one" };
+		maxTimeout = 31L;
+		context = "some context";
+
+		semaphore = new WriteReadSemaphoreImpl(mockCountingSemaphore, maxNumberOfReaders);
+	}
+
+	@Test
+	public void testSemaphoreWithNullCounting() {
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			new WriteReadSemaphoreImpl(null, maxNumberOfReaders);
+		}).getMessage();
+		assertEquals("CountingSemaphore cannot be null", message);
+	}
+
+	@Test
+	public void testGetReadLockProviderWithNullRequest() {
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			semaphore.getReadLock(null);
+		}).getMessage();
+		assertEquals("Request cannot be null", message);
+	}
+	
+	@Test
+	public void testGetWriteLockProviderWithNullRequest() {
+		String message = assertThrows(IllegalArgumentException.class, () -> {
+			// call under test
+			semaphore.getWriteLock(null);
+		}).getMessage();
+		assertEquals("Request cannot be null", message);
+	}
+
+	@Test
+	public void testGetReadLockProvider() throws Exception {
+		
+		when(mockCountingSemaphore.attemptToAcquireLock(any(), anyLong(), anyInt(), any())).thenReturn(
+				Optional.of("tokenOne"));
+		when(mockCountingSemaphore.getFirstUnexpiredLockContext(any())).thenReturn(Optional.empty());
+		when(mockCallback.getLockTimeoutSeconds()).thenReturn(maxTimeout);
+
+		// call under test
+		try (ReadLock provider = semaphore
```

Comment:
you could mock the actual lock to just test the ogic of the writereadsemaphore (e.g. all on the lock.acquire and close)